BS-14438 As Dave, I want to be able to automatically check mandatory

azara opened the pull request:
attributes when generating contract input from Business Data

add a column to display if the attribute is mandatory or not
add mandatory attributes button
add method to check mandatory attributes
add a warning validation

apupier commented on bundles/plugins/org.bonitasoft.studio.contract/messages.properties:
```diff
@@ -79,3 +79,5 @@ generatedScriptPreviewTitle=Generated script preview
 selectAll=Select all
 deselectAll=Deselect all
+selectMandatories=Mandatory attributes
+mandatoryFieldsNotSelectedWarning=Mandatory field(s) {0}may be selected. If not, manual edition of the initialization script will be needed.
```
missing space after {0}

apupier commented on bundles/plugins/org.bonitasoft.studio.contract/src/org/bonitasoft/studio/contract/ui/wizard/CreateContractInputFromBusinessObjectWizardPage.java:
```diff
@@ -308,25 +315,71 @@ protected MultiValidator createEmptySelectionMultivalidator(final IViewerObserva
+            final List<FieldToContractInputMapping> mappings, final IObservableSet checkedElements) {
+        for (final FieldToContractInputMapping mapping : mappings) {
+            if (!checkedElements.contains(mapping) && !mapping.isGenerated() && !mapping.getField().isNullable()) {
+                sb.append(mapping.getField().getName() + ", ");
```
better to use two times append instead of +

apupier commented on bundles/plugins/org.bonitasoft.studio.contract/src/org/bonitasoft/studio/contract/ui/wizard/labelProvider/MandatoryColumnLabelProvider.java:
```diff
@@ -0,0 +1,38 @@
+    public String getText(final Object element) {
+        if (element instanceof FieldToContractInputMapping) {
+            final FieldToContractInputMapping mapping = (FieldToContractInputMapping) element;
+            return new Boolean(!mapping.getField().isNullable()).toString();
```
Boolean.valueOf to avoid  object instanciation

apupier commented on bundles/plugins/org.bonitasoft.studio.contract/src/org/bonitasoft/studio/contract/ui/wizard/CreateContractInputFromBusinessObjectWizardPage.java:
```diff
@@ -308,25 +315,72 @@ protected MultiValidator createEmptySelectionMultivalidator(final IViewerObserva
+                        checkedElements.add(mapping);
+                        checkAllMappings(checkedElements, mapping.getChildren());
+                        generateAllMappings(mapping.getChildren(), true);
+                        if (mapping.getParent() != null && !checkedElements.contains(mapping.getParent())) {
```
mapping.getParent can be extracted as avriable

apupier commented on bundles/plugins/org.bonitasoft.studio.contract/src/org/bonitasoft/studio/contract/ui/wizard/CreateContractInputFromBusinessObjectWizardPage.java:
```diff
@@ -308,25 +315,72 @@ protected MultiValidator createEmptySelectionMultivalidator(final IViewerObserva
+                for (final FieldToContractInputMapping mapping : mappings) {
+                    if (!mapping.getField().isNullable() && !mapping.isGenerated()) {
+                        checkedElements.add(mapping);
+                        checkAllMappings(checkedElements, mapping.getChildren());
```
mapping.getCHildren can be extracted as a variable

apupier commented:
When an attribute is not mandatory and have manadatory attributes, the warning about mandatory fields should nbe displayed only if this attribute is checked. I tshoudl nto; be displayed if attribute is not checked

apupier commented:
Going back and forth to select another business data doesn't refresh the validation about missing manadatory attributes

apupier commented:
- unselect all
- click on mandatory attributes
- click "Next" to preview
--> the script is not the expected one

azara commented:
the defect of script generation has also been corrected.

apupier commented:
the issue
"When an attribute is not mandatory and have manadatory attributes, the warning about mandatory fields should nbe displayed only if this attribute is checked. I tshoudl nto; be displayed if attribute is not checked" still appears

apupier commented:
he issue
"When an attribute is not mandatory and have manadatory attributes, the warning about mandatory fields should nbe displayed only if this attribute is checked. I tshoudl nto; be displayed if attribute is not checked" still appears
![image](https://cloud.githubusercontent.com/assets/1105127/10424553/44b196d2-70d0-11e5-99c5-da59947abe2c.png)

azara commented:
Strange, you don't seem to be up to date 
![image](https://cloud.githubusercontent.com/assets/5079838/10424663/4eab391c-70d1-11e5-9f3d-4a3093e3ee0a.png)

apupier commented:
OK for validation, seems that I wasn't on the correct branch (maybe due to rebase and so on force history overwrite)

apupier commented:
cannot generate when a mandatory field is not checked

apupier commented:
NPE when clicking on Deslect all: !ENTRY org.eclipse.ui 4 0 2015-10-19 15:18:42.654
!MESSAGE Unhandled event loop exception
!STACK 0
java.lang.NullPointerException
	at org.bonitasoft.studio.contract.ui.wizard.CreateContractInputFromBusinessObjectWizardPage$8.widgetSelected(CreateContractInputFromBusinessObjectWizardPage.java:402)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:248)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4353)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1061)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4172)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3761)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:832)
	at org.eclipse.jface.window.Window.open(Window.java:808)
	at org.bonitasoft.studio.contract.ui.property.ContractPropertySection.openAddInputWizardDialog(ContractPropertySection.java:261)
	at org.bonitasoft.studio.contract.ui.property.ContractPropertySection$1.widgetSelected(ContractPropertySection.java:244)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:248)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4353)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1061)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4172)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3761)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$9.run(PartRenderingEngine.java:1151)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:332)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1032)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:148)
	at org.eclipse.ui.internal.Workbench$5.run(Workbench.java:636)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:332)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:579)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
	at org.bonitasoft.studio.application.BonitaStudioApplication.createAndRunWorkbench(BonitaStudioApplication.java:93)
	at org.bonitasoft.studio.application.BonitaStudioApplication.start(BonitaStudioApplication.java:88)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:380)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:235)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:648)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:603)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1465)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1438)

apupier commented:
NPE was due to the move of buttons from left to right and not the same order of initialization of CheckedElements Observable --> will place them back to the right for now

apupier commented:
when clicking "Finish & Add" then "Cancel", the "preview" button becomes disabled

apupier commented:
On task, the warning message for mandatory attributes not selected should not be the same as it can be just an update and not an initialization

apupier commented on bundles/plugins/org.bonitasoft.studio.contract/messages.properties:
```diff
@@ -58,25 +58,28 @@ atLeastOneAttributeShouldBeSelectedError=At least one attribute should be checke
 deselectAll=Deselect all
 preview=Preview >
+selectMandatories=Mandatory attributes
+mandatoryFieldsNotSelectedWarning=Mandatory attribute(s) {0} is/are not selected. Either select them or initialize them manually.
```
select it/them ?

apupier commented on bundles/plugins/org.bonitasoft.studio.contract/src/org/bonitasoft/studio/contract/ui/wizard/CreateContractInputFromBusinessObjectWizardPage.java:
```diff
@@ -308,25 +320,85 @@ protected MultiValidator createEmptySelectionMultivalidator(final IViewerObserva
             protected IStatus validate() {
                 if (checkedElements.isEmpty()) {
                     return ValidationStatus.error(Messages.atLeastOneAttributeShouldBeSelectedError);
+                } else {
```
The anonymous class starts to be complicated, aybe extracted it to a specific class will simplify code and help to test easily

apupier commented on bundles/plugins/org.bonitasoft.studio.contract/messages.properties:
```diff
@@ -58,25 +58,28 @@ atLeastOneAttributeShouldBeSelectedError=At least one attribute should be checke
 autoGeneratedOperationStepButton=Auto-generate storage operations
 manuallyDefinedOperationStepButton=No, thanks. I'll manually define how to use the contract.
-autoGeneratedOperationProcessButton=Auto-generate the initialization script
+autoGeneratedOperationProcessButton=Auto-generate the initialization script.Click on preview button to check its content.
```
Missing space before "Click on preview"

apupier commented:
Validation is not refreshed when going back and forth
- have two business data using mandatory attributes
- "Add from data.." choose the first one and uncheck some mandatory attributes
- click backl to select another business data
- click next
--> the valiation is still mentioning the warning on mandatory attributes not checked from the first businees data

apupier commented:
The list of mandatory attributes in the warnign message finishes with a "," which leads to a non-correct syntacxic sentence
sample:
"Mandatory attribute(s) attribute2.attribute22m, attribute3m.attribute4m,  is/are not selected. Either select it/them or create storage and update operations."

apupier commented:
On a task, therer is still the popup stating "Incomplete initialization of the business variable" although it can be an update operation

azara commented:
La virgule marque une courte pause dans la lecture sans cependant que l'intonation change. 

La virgule s'emploie : 

-> Dans une énumération, pour séparer des mots, des groupes de mots de même nature ou des propositions juxtaposées.
Elle monte, elle descend, elle n'arrête pas de bouger !
Les lions, les girafes, les zèbres, vivent tous trois dans la savane. 

Unless rules are different between french and english the "," is correct at the end of enumeration.

apupier commented:
ok for the comma, learned a thing.
Does it apply for enumeration of one element only?

azara commented:
I removed the extra "," when there is only one mandatory attribute unchecked.
I have extracted a class for this  multivalidator and I have changed the text of the last pop up when user is on a task.
The problem of validation refresh when returning on data choice is all ready in maintenance backlog. I tried to correct it but I didn't find solution.

apupier commented:
in some simple cases, the warning validation of missing mandatory attributes is not displayed:
![image](https://cloud.githubusercontent.com/assets/1105127/11180111/7a8146d8-8c58-11e5-95f2-889df4c83c36.png)

apupier commented on bundles/plugins/org.bonitasoft.studio.contract/src/org/bonitasoft/studio/contract/ui/wizard/EmptySelectionMultivalidator.java:
```diff
@@ -0,0 +1,83 @@
+                }
+                sb.append(mapping.getField().getName());
+                sb.append(", ");
+
```
useless emtpy line

apupier commented on bundles/plugins/org.bonitasoft.studio.contract/src/org/bonitasoft/studio/contract/ui/wizard/EmptySelectionMultivalidator.java:
```diff
@@ -0,0 +1,83 @@
+/**
+ * @author aurelie
+ */
+public class EmptySelectionMultivalidator extends MultiValidator {
```
there is no unit test for this class
